lib/dry/auto_inject/strategies/kwargs.rb

```diff
@@ -24,59 +25,74 @@ def define_new
         end
 
         def define_initialize(klass)
-          super_method = Dry::AutoInject.super_method(klass, :initialize)
+          super_parameters = Dry::AutoInject.super_parameters(klass, :initialize).first
 
-          if super_method.nil? || super_method.parameters.empty?
-            define_initialize_with_keywords
+          if super_parameters.splat? || super_parameters.sequential_arguments?
+            define_initialize_with_splat(super_parameters)
           else
-            define_initialize_with_splat(super_method)
+            define_initialize_with_keywords(super_parameters)
           end
 
           self
         end
 
-        def define_initialize_with_keywords
-          initialize_params = dependency_map.names.map { |name| "#{name}: nil" }.join(", ")
+        def define_initialize_with_keywords(super_parameters)
+          assign_dependencies = method(:assign_dependencies)
+          slice_kwargs = method(:slice_kwargs)
 
-          instance_mod.class_eval <<-RUBY, __FILE__, __LINE__ + 1
-            def initialize(#{initialize_params})
-              #{dependency_map.names.map { |name| "@#{name} = #{name} unless #{name}.nil? && instance_variable_defined?(:'@#{name}')" }.join("\n")}
-              super()
-            end
-          RUBY
+          instance_mod.class_exec(dependency_map) do |dependency_map|
+            define_method :initialize do |**kwargs|
```

Comment:
exactly this, otherwise we would come up with two `define_initialize_with_keywords` with unclear differences between them. Now we at least have east-to-grasp logic there

Reply:
Yep, I think clarity of internal implementation is probably the better thing to optimise for here 👍